File: 4.cpp
#include <iostream>
#include <bitset>
#include <vector>
#include <string>

class NetAddress {
private:
    uint32_t address{ 0 };
    uint32_t netmask{ 0 };

    std::vector<uint8_t> split(std::string s, char delim) {
        std::vector<uint8_t> result;
        size_t pos = 0;
        while ((pos = s.find(delim)) != std::string::npos) {
            result.push_back(stoi(s.substr(0, pos)));
            s.erase(0, pos + 1);
        }
        result.push_back(stoi(s));
        return result;
    }

public:
    NetAddress(std::string ip, std::string mask) {

        auto ips = split(ip, '.');
        for (int i = 0; i < 4; ++i) {
            address |= (ips[i] << (i * 8));
        }

        auto masks = split(mask, '.');
        for (int i = 0; i < 4; ++i) {
            netmask |= (masks[i] << (i * 8));
        }
    }

    NetAddress(uint32_t address, uint32_t netmask)
        : address(address), netmask(netmask) {}

    NetAddress operator&(const NetAddress& other) const {
        return NetAddress(address & other.address, netmask & other.netmask);
    }

    bool operator==(const NetAddress& other) const {
        return address == other.address && netmask == other.netmask;
    }

    friend std::ostream& operator<<(std::ostream& os, const NetAddress& na);


};

std::ostream& operator<<(std::ostream& os, const NetAddress& na) {
    os << std::bitset<8>(na.address >> 24) << "."
        << std::bitset<8>(na.address >> 16) << "."
        << std::bitset<8>(na.address >> 8) << "."
        << std::bitset<8>(na.address) << "/"
        << std::bitset<32>(na.netmask);
    return os;
}

int main() {
    NetAddress ip("192.168.0.1", "255.255.255.0");
    NetAddress subnet = ip & ip;

    std::cout << "IP: " << ip << std::endl;
    std::cout << "Subnet: " << subnet << std::endl;

    std::cout << "('192.168.0.1', '255.255.255.0') == ('192.168.0.1', '255.255.255.0') ? -> " << (ip == ip) << std::endl;

    return 0;
}

File: 4.py

class NetAddress:
    def __init__(self, ip, mask):
        self.address = self._parse_ip(ip)
        self.netmask = self._parse_mask(mask)

    def __and__(self, other):
        return NetAddress(self.address & other.address, 
                          self.netmask & other.netmask)
                          
    def __eq__(self, other):
        return self.address == other.address and self.netmask == other.netmask

    def _parse_ip(self, ip):
        return sum(int(x) << (i * 8) for i, x in enumerate(ip.split('.')))
    
    def _parse_mask(self, mask):
        return sum(int(x) << (i * 8) for i, x in enumerate(mask.split('.')))
        
    def __str__(self):
        return f"{self._int_to_ip(self.address)}/{self._int_to_mask(self.netmask)}"

    def _int_to_ip(self, addr):
        return '.'.join(str(addr >> (i * 8) & 0xff) for i in range(3,-1,-1))

    def _int_to_mask(self, mask):
        return '.'.join(str(mask >> (i * 8) & 0xff) for i in range(3,-1,-1))

    def get_network_address(self):
        network_addr = self.address & self.netmask
        return f"{network_addr&0xFF}.{(network_addr>>8)&0xFF}.{(network_addr>>16)&0xFF}.{(network_addr>>24)&0xFF}"

ip = NetAddress("192.168.0.1", "255.255.255.0")
subnet = ip.get_network_address()

print(f"IP: {ip}")
print(f"Subnet: {subnet}")
print(f"ip == ip ? -> {(ip==ip)}")
